### DISCUSSION TITLE
getStaticPaths and getStaticProps work in development but the prop is undefined in build time

### DISCUSSION THREAD

**Discussion by yasserelsaid** [vercel/next.js#16712] [Help]
I am making an eCommerce platform using Next.js. In the product page, where I use dynamic routes, I used getStaticProps to fetch my product and getStaticPaths to generate the paths to the products that should be statically generated. This works perfectly in development mode but the project does not build. The product prop which is passed from getStaticProps to the page component is undefined in build time.

This is the error I get when trying to build:

> Error occurred prerendering page "/product/[id]". Read more: https://err.sh/next.js/prerender-error
> TypeError: Cannot destructure property 'name' of 'product' as it is undefined.

Here is my code:
```

function ProductPage({ product}) {
// product is defined in development but undefined when trying to build
const {
name,
price,
} = product;
// The rest of the component
}

export const getStaticPaths = async () => {
  connectDb();

  const products = await Product.find();
  const paths = products.map(product => ({
    params: { id: product._id.toString() },
  }));
  return {
    paths,
    fallback: true,
  };
};

export const getStaticProps = async ({ params }) => {
  connectDb();
  const { id } = params;
  try {
    // getting the product
    const product = await Product.findOne({ _id: id });
    return {
      props: {
        product: JSON.parse(JSON.stringify(product)),
      },
    };
  } catch (err) {
    console.log(err);
    return {
      props: {
        product: {},
      },
    };
  }
};

export default ProductPage;

```

**Comment by jamesmosier** [comment]
If you log out `product` (`console.log(product)`) in `getStaticProps` during a build, what is the output of that log?

**Reply by yasserelsaid** [reply]
it logs 'undefined' and it seems that the page component function (ProductPage) runs before getStaticProps and getStaticPaths of the same page, because when I tried to log a message in getStaticPaths, the message was shown after the console.log from the ProductPage function. 

I feel this has something to do with how my _app page is structured. I fetch the user in MyApp.getInitialProps and pass it down to the page component. I also redirect from protected routes in MyApp.getInitialProps
Here is the code:
```
function MyApp({ Component, pageProps, appProps }) {
  return (
    <React.Fragment>
      <Header
        links={
          <HeaderLinks user={appProps.user} dropdownHoverColor='primary' />
        }
      />
        <Component {...pageProps} {...appProps} />
      <MyFooter />
    </React.Fragment>
  );
}

MyApp.getInitialProps = async ({ Component, ctx }) => {
  const { token } = parseCookies(ctx);
  let appProps = {};
  if (!token) {
    const isProtectedRoute =
      ctx.pathname === '/account' || ctx.pathname === '/create';
    if (isProtectedRoute) {
      redirectUser(ctx, '/login');
    }
  } else {
    try {
      const payload = { headers: { Authorization: token } };
      const url = `${baseUrl}/api/account`;
      const response = await axios.get(url, payload);
      const user = response.data;
      // if authenticated but not admin or root then redirect from the create page
      const isRoot = user.role === 'root';
      const isAdmin = user.role === 'admin';
      const isNotPermitted = !(isRoot || isAdmin) && ctx.pathname === '/create';
      if (isNotPermitted) {
        redirectUser(ctx, '/');
      }
      appProps.user = user;
    } catch (error) {
      console.error('Error getting current user', error);
      // delete invalid tokens
      destroyCookie(ctx, 'token');
      // redirect to login
      redirectUser(ctx, '/login');
    }
  }
  return { appProps };
};

export default MyApp;
```

**Reply by jamesmosier** [reply]
`MyApp.getInitialProps` will only be run on the client and server during runtime, so any logic in there will not be used during `getStaticProps` or `getStaticPaths`.

Aside from that, it seems like the issue would be within `Product.findOne({ _id: id });` or `connectDb`. But i'm not sure what the logic in there looks like so it is hard to say.

**Comment by thepors** [comment]
@yasserelsaid I am facing same problem. Did you get any solution?

**Comment by zakariamofaddel** [comment]
I was having the same exact issue when creating static paths for all the blog posts in our CMS, it worked in development but not when building.
In a better reading session in the NextJs docs I realised that having the fallback to true doesn't make sense since I'm fetching all the blog posts permalinks and for anything else that doesn't match my blog posts permalink, I want to serve a 404.  
So for me it solved it to simply turn the fallback key's value to false:
````
export const getStaticPaths = async () => {

  (fetching data...)

  return {
    paths,
    fallback: false,
  };
};
````

**Comment by pveyes** [comment]
From next.js documentation https://nextjs.org/docs/basic-features/data-fetching#fallback-pages, **using `fallback: true` will make props to be empty** in some cases, this is why you have to provide fallback page by checking `router.isFallback` value

Alternatively you can use `fallback: 'blocking'`, if you want to generate the rest of the page in runtime, or `fallback: false` if you want to generate 404

**Comment by masonbrothers** [comment]
I found that if it meets your use case, you can have some kind of fallback at the top of the page. I think there are multiple renders, and `undefined` is passed to page props on the first render. On the second, the the actual value is passed.

You might be able to do this if it suits your use case:
```javascript
function ProductPage({ product}) {
  if (typeof product === "undefined") return "Loading..."
...
```
